### PR TITLE
dialects: (arith) add SignlessIntegerBinaryOperation canonicalization

### PR DESCRIPTION
**Pull request by alexarice**
Got bored of implementing various arith canonicalizations that all looked the same, so I've done a bunch of them generically in one go.

I've missed out a fair few methods (shifting operations in particular stand out) where I wasn't sure if the python operation agreed with the mlir operation.

### CODE REVIEW

**Review by superlopuh — APPROVED** [review]

File: xdsl/dialects/arith.py [hunk 1]
```diff
@@ -190,6 +191,18 @@ class SignlessIntegerBinaryOperation(IRDLOperation, abc.ABC):
...
+    def py_operation(lhs: int, rhs: int) -> int | None:
+        return None
+
+    @staticmethod
+    def is_right_zero(attr: AnyIntegerAttr) -> bool:
+        return False
+
+    @staticmethod
+    def is_right_unit(attr: AnyIntegerAttr) -> bool:
+        return False
```
Comment (math-fehr) [inline xdsl/dialects/arith.py:204]: Can you add documentation on that function?
Why is there a `right` in the name? It doesn't seem that it takes the left/right side into account?

File: xdsl/dialects/arith.py [hunk 2]
```diff
@@ -395,19 +425,27 @@ def infer_overflow_type(input_type: Attribute) -> Attribute:
...
+        SignlessIntegerBinaryOperationHasCanonicalizationPatternsTrait(),
+    )
+
+    @staticmethod
+    def py_operation(lhs: int, rhs: int) -> int | None:
+        return lhs * rhs
+
+    @staticmethod
+    def is_right_unit(attr: AnyIntegerAttr) -> bool:
+        return attr == IntegerAttr(1, attr.type)
```
Comment (math-fehr) [inline xdsl/dialects/arith.py:444]: Why is this one taking the type into account, but the one in Addi just checks the value?

**Review by math-fehr — APPROVED** [review]
Nice!

**Comment by math-fehr** [comment]
Btw, at some point we should define an interface for `is_zero` and `is_unit`.
Not the scope of this PR though

**Comment by superlopuh** [comment]
`is_unit` is relative to the operation, though

File: xdsl/dialects/arith.py (hunk 1 above)
Comment (alexarice) [reply xdsl/dialects/arith.py:204]: `i` being a right unit means that `x . i = x` but we don't necessarily have `i . x = x`. See the shift operations for example

File: xdsl/dialects/arith.py (hunk 2 above)
Comment (alexarice) [reply xdsl/dialects/arith.py:444]: The one here is just to work around that the unit for `i1` multiplication is `-1` and not `1`. This doesn't matter in the 0 case which `Addi` has

Comment (alexarice) [reply xdsl/dialects/arith.py:444]: Thinking about it I'm not 100% sure this will work for `si1`

**Comment by alexarice** [comment]
`is_right_zero` is also relative to the operation.

**Comment by alexarice** [comment]
I've realised that being a zero for a function might not be common terminology for non-mathematicians. Would it be preferable to rename this to `is_right_absorbing` or `is_right_annihilating` instead?

**Comment by superlopuh** [comment]
Both of those seem more confusing 😅

File: xdsl/dialects/arith.py [hunk 3]
```diff
@@ -190,6 +191,29 @@ class SignlessIntegerBinaryOperation(IRDLOperation, abc.ABC):
...
+        Note that this depends on the operation and does *not* imply that
+        attr.value.data == 0
+        """
+        return False
+
+    @staticmethod
+    def is_right_unit(attr: AnyIntegerAttr) -> bool:
+        """
+        Return True only when 'attr' is a right unit/identity for the operation
+        https://en.wikipedia.org/wiki/Identity_element
```
Comment (superlopuh) [inline xdsl/dialects/arith.py:213]: ooh, if there are wikipedia pages with those titles then maybe using those for function names is fair game

**Comment by alexarice** [comment]
This suffers from a similar problem to assert #3585 . I'll try to work out the proper overflow semantics

**Comment by alexarice** [comment]
I've rebased, added an `i1` test, and truncated the result of constant propagation to fix this case and keep in line with what mlir does.

File: xdsl/dialects/arith.py (hunk 2 above)
Comment (alexarice) [reply xdsl/dialects/arith.py:460]: @superlopuh is there a better way to do this check?

File: xdsl/dialects/arith.py [hunk 4]
```diff
@@ -166,13 +167,18 @@ def __init__(
...
+        truncate_bits: bool = False,
     ) -> ConstantOp:
         if isinstance(value_type, int):
             value_type = IntegerType(value_type)
         return ConstantOp.create(
             result_types=[value_type],
-            properties={"value": IntegerAttr(value, value_type)},
+            properties={
+                "value": IntegerAttr(value, value_type, truncate_bits=truncate_bits)
+            },
```
Comment (superlopuh) [inline xdsl/dialects/arith.py:181]: how do you feel about a quick separate PR for this guy?

Comment (alexarice) [reply xdsl/dialects/arith.py:181]: #3635

**Comment by alexarice** [comment]
Should be ready for review again

**Comment by alexarice** [comment]
@superlopuh @compor @math-fehr I realise you've all approved but just want to check that you're all happy for this to be merged as I think a couple of things have changed since review

**Comment by compor** [comment]
> @superlopuh @compor @math-fehr I realise you've all approved but just want to check that you're all happy for this to be merged as I think a couple of things have changed since review

having another look now

File: xdsl/dialects/arith.py [hunk 5]
```diff
@@ -195,6 +196,29 @@ class SignlessIntegerBinaryOperation(IRDLOperation, abc.ABC):
 
     assembly_format = "$lhs `,` $rhs attr-dict `:` type($result)"
 
+    @staticmethod
+    def py_operation(lhs: int, rhs: int) -> int | None:
```
Comment (compor) [inline xdsl/dialects/arith.py:200]: Could you please add a simple doc string on what this does/is for? I find the name very unintuitive, although I understand it's hard (this looks like an `eval` as we have for affine maps?)

File: xdsl/dialects/arith.py (hunk 2 above)
Comment (compor) [reply xdsl/dialects/arith.py:460]: this is what we seem to be doing in other places too

**Review by compor — APPROVED** [review]

File: xdsl/dialects/arith.py (hunk 2 above)
Comment (superlopuh) [reply xdsl/dialects/arith.py:460]: I can't think of one

**Review by superlopuh — APPROVED** [review]

File: xdsl/dialects/arith.py (hunk 5 above)
Comment (superlopuh) [reply xdsl/dialects/arith.py:200]: ah wait a second could this be used in the interpreter also?

Comment (alexarice) [reply xdsl/dialects/arith.py:200]: Have added a docstring. Essentially this should return an integer when we know a canonicalization can happen. Unsure if this can/should be used in the interpreter
